essential_reading/DataSources/Data_S3.md

```diff
@@ -0,0 +1,223 @@
+'type': 'file',
+'size': 32357,
+'name': 'noaa-nwm-retrospective-2-1-zarr-pds/index.html'}
+```
+
+Low-level operations with `fsspec` are designed to behave similarly to the 'native' file operations in Python.
+
+```python
+# Use open() to get something that behaves like a file handle for
+# low-level Python read/write operations:
+with fs.open('s3://noaa-nwm-retrospective-2-1-zarr-pds/index.html') as f:
+    # print first 5 lines...
+    for i in range(0, 5):
+        line = f.readline()
+        print(line)
+```
+
+Output:
+
+```text
+b'<!DOCTYPE html>\r\n'
+b'\r\n'
+b'<!--\r\n'
+b'Copyright 2014-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.\r\n'
+b'\r\n'
+```
+
+However -- most of the time, we don't need such low level access.
+A more general, and useful, mechanism is to '_map_' an S3 object to a _virtual file_ using `get_mapper()`.
+That _map_ object can then be treated as if it is a local file -- the `fsspec` mapping mechanism handles
+all of the translation to interact with S3.
+
+```python
+m = fs.get_mapper('s3://noaa-nwm-retrospective-2-1-zarr-pds/chrtout.zarr')
+# 'm' looks like a file as far as Python is concerned.
+# It is now a stand-in for those functions that require a file as argument.
+g = zarr.convenience.open_consolidated(m) # read zarr metadata for named file.
+print(g.tree())
+```
+
+Output:
+
+```text
+/
+ ├── crs () |S1
+ ├── elevation (2776738,) float32
+ ├── feature_id (2776738,) int32
+ ├── gage_id (2776738,) |S15
+ ├── latitude (2776738,) float32
+ ├── longitude (2776738,) float32
+ ├── order (2776738,) int32
+ ├── streamflow (367439, 2776738) int32
+ ├── time (367439,) int64
+ └── velocity (367439, 2776738) int32
+```
+
+:::{warning}
+
+We are deliberately demonstrating using commands that read **only** the zarr metadata from the
+object storage -- **not the full data set**.
+
+The above example is a very, _very_, **very** large dataset, which you don't want to load over the network to your desktop.
+Execute full data read operations only if this notebook is being hosted and run out of the same AWS center where the data lives.
+
+:::
+
+:::{sidebar}
+
+The good news about some of the larger science-oriented libraries (xarray, dask, pandas, zarr, etc), is that
+they can automatically handle the `fsspec` operations for you **IF YOUR ACCESS IS ANONYMOUS**.
+This is a convenience, but is a special case for read-only data where `anon=True` and `requestor_pays=False`.
+
+Note that this is a feature of
+[specific libraries](https://filesystem-spec.readthedocs.io/en/latest/#who-uses-fsspec),
+and doesn't work everywhere.
+
+Because it isn't universally available, and only applies to anonymous reads, examples in HyTEST will
+always explicitly plumb `fsspec` 'longhand' using `get_mapper()`. You may see example code elsewhere
+that takes the shortcut if it is available.
+
+:::
+
+
+## Credentialed Access
+
+For most data storage within the HyTEST workflows, access will not be anonymous.
```

Comment:
I don't want to assume this will always be true!
```suggestion
For some data storage within the HyTEST workflows, access will not be anonymous.
```